pkgs/stdenv/generic/setup.sh

```diff
@@ -409,8 +409,10 @@ concatTo() {
                         # Reproduces https://github.com/NixOS/nixpkgs/pull/318614/files#diff-7c7ca80928136cfc73a02d5b28350bd900e331d6d304857053ffc9f7beaad576L359
                         targetref+=( ${nameref+"${nameref[@]}"} )
                     else
+                        set -o noglob
                         # shellcheck disable=SC2206
                         targetref+=( ${nameref-} )
+                        set +o noglob
```

Comment:
You can use `local -` to make shell options local to the function call. Technically, this requires Bash 4.4, but see #340765.

https://www.gnu.org/software/bash/manual/html_node/Bash-Builtins.html#index-local:~:text=If%20name%20is%20,local%20is%20invoked